File: src/moltblock/entity.py
"""Entity: minimal runnable loop — task in -> graph (Generator/Critic/Judge) -> Verifier -> artifact out."""

from pathlib import Path

from .agents import run_critic, run_generator, run_judge
from .config import ModelBinding, default_code_entity_bindings
from .gateway import LLMGateway
from .graph_runner import GraphRunner
from .graph_schema import AgentGraph
from .memory import WorkingMemory
from .persistence import Store, hash_memory, record_outcome
from .verifier import run_verifier


class CodeEntity:
    """
    Code Entity: Generator -> Critic -> Judge -> Verifier.
    Uses working memory and per-role LLM gateways.
    """

    def __init__(self, bindings: dict[str, ModelBinding] | None = None) -> None:
        bindings = bindings or default_code_entity_bindings()
        self._gateways = {
            "generator": LLMGateway(bindings["generator"]),
            "critic": LLMGateway(bindings["critic"]),
            "judge": LLMGateway(bindings["judge"]),
        }

    def run(
        self,
        task: str,
        test_code: str | None = None,
        store: Store | None = None,
        entity_version: str = "0.2.0",
        write_checkpoint_after: bool = False,
    ) -> WorkingMemory:
        """
        One full loop: task in -> Generator -> Critic -> Judge -> Verifier -> gating.
        If store is provided and verification passed: admit to verified memory; optionally write checkpoint.
        Returns working memory with authoritative_artifact set only if verification passed.
        """
        import time
        t0 = time.perf_counter()
        memory = WorkingMemory()
        memory.set_task(task)

        if store:
            recent = store.get_recent_verified(5)
            parts = []
            for e in recent:
                if e.get("content_preview"):
                    parts.append(e["content_preview"][:500])
                elif e.get("summary"):
                    parts.append(e["summary"])
            memory.long_term_context = "\n---\n".join(parts) if parts else ""

        run_generator(self._gateways["generator"], memory, store)
        run_critic(self._gateways["critic"], memory, store)
        run_judge(self._gateways["judge"], memory, store)
        run_verifier(memory, test_code=test_code)

        if store:
            record_outcome(store, memory.verification_passed, time.perf_counter() - t0, task[:100])
        if store and memory.verification_passed and memory.authoritative_artifact:
            artifact_ref = f"artifact_{int(time.time() * 1000)}"
            store.add_verified(
                artifact_ref,
                summary=f"Verified artifact ({len(memory.authoritative_artifact)} chars)",
                content_preview=memory.authoritative_artifact[:2000],
            )
            if write_checkpoint_after:
                graph_hash = "fixed-code-entity"
                refs = [artifact_ref]
                mem_hash = hash_memory(refs)
                store.write_checkpoint(entity_version, graph_hash, mem_hash, refs)

        return memory


def load_entity_with_graph(
    graph_path: str | Path,
    bindings: dict[str, ModelBinding] | None = None,
) -> GraphRunner:
    """Load an Entity from a declarative graph (JSON/YAML). Returns a GraphRunner."""
    graph = AgentGraph.load(graph_path)
    return GraphRunner(graph, bindings=bindings)

File: src/entity.ts
/**
 * Entity: minimal runnable loop — task in -> graph (Generator/Critic/Judge) -> Verifier -> artifact out.
 */

import { runCritic, runGenerator, runJudge } from "./agents.js";
import { defaultCodeEntityBindings, type ModelBinding } from "./config.js";
import { LLMGateway } from "./gateway.js";
import { GraphRunner } from "./graph-runner.js";
import { AgentGraph } from "./graph-schema.js";
import { WorkingMemory } from "./memory.js";
import { Store, hashMemory, recordOutcome } from "./persistence.js";
import { runVerifier } from "./verifier.js";

/**
 * Code Entity: Generator -> Critic -> Judge -> Verifier.
 * Uses working memory and per-role LLM gateways.
 */
export class CodeEntity {
  private gateways: Record<string, LLMGateway>;

  constructor(bindings?: Record<string, ModelBinding>) {
    const resolvedBindings = bindings ?? defaultCodeEntityBindings();
    this.gateways = {
      generator: new LLMGateway(resolvedBindings["generator"]!),
      critic: new LLMGateway(resolvedBindings["critic"]!),
      judge: new LLMGateway(resolvedBindings["judge"]!),
    };
  }

  /**
   * One full loop: task in -> Generator -> Critic -> Judge -> Verifier -> gating.
   * If store is provided and verification passed: admit to verified memory; optionally write checkpoint.
   * Returns working memory with authoritative_artifact set only if verification passed.
   */
  async run(
    task: string,
    options: {
      testCode?: string;
      store?: Store;
      entityVersion?: string;
      writeCheckpointAfter?: boolean;
    } = {}
  ): Promise<WorkingMemory> {
    const {
      testCode,
      store,
      entityVersion = "0.2.0",
      writeCheckpointAfter = false,
    } = options;

    const t0 = performance.now();
    const memory = new WorkingMemory();
    memory.setTask(task);

    // Inject long-term context from verified memory
    if (store) {
      const recent = store.getRecentVerified(5);
      const parts: string[] = [];
      for (const e of recent) {
        if (e.content_preview) {
          parts.push(e.content_preview.slice(0, 500));
        } else if (e.summary) {
          parts.push(e.summary);
        }
      }
      memory.longTermContext = parts.length > 0 ? parts.join("\n---\n") : "";
    }

    // Run the agent pipeline
    await runGenerator(this.gateways["generator"]!, memory, store ?? null);
    await runCritic(this.gateways["critic"]!, memory, store ?? null);
    await runJudge(this.gateways["judge"]!, memory, store ?? null);
    await runVerifier(memory, testCode);

    // Record outcome and persist if verification passed
    const latencySec = (performance.now() - t0) / 1000;
    if (store) {
      recordOutcome(store, memory.verificationPassed, latencySec, task.slice(0, 100));
    }

    if (store && memory.verificationPassed && memory.authoritativeArtifact) {
      const artifactRef = `artifact_${Date.now()}`;
      store.addVerified(
        artifactRef,
        `Verified artifact (${memory.authoritativeArtifact.length} chars)`,
        memory.authoritativeArtifact.slice(0, 2000)
      );

      if (writeCheckpointAfter) {
        const graphHash = "fixed-code-entity";
        const refs = [artifactRef];
        const memHash = hashMemory(refs);
        store.writeCheckpoint(entityVersion, graphHash, memHash, refs);
      }
    }

    return memory;
  }
}

/**
 * Load an Entity from a declarative graph (JSON/YAML). Returns a GraphRunner.
 */
export function loadEntityWithGraph(
  graphPath: string,
  bindings?: Record<string, ModelBinding>
): GraphRunner {
  const graph = AgentGraph.load(graphPath);
  return new GraphRunner(graph, bindings);
}
